### PR TITLE
Doc classifier

### PR DESCRIPTION
**Pull request by ankrgyl**
Replace the hacked image classification pipeline with a new `DocumentClassificationPipeline`, which is loosely based on the DocumentQuestionAnsweringPipeline` and `TextClassificationPipeline`.

### CODE REVIEW

**Comment by jeffsrobertson** [comment]
Nice! Will start digging in now 👍

**Comment by jeffsrobertson** [comment]
There are two instances of `"image-classification"` which need to be replaced by `"document-classification"`in `src/docquery/cmd/scan.py`. Specifically, [here](https://github.com/impira/docquery/blob/doc-classifier/src/docquery/cmd/scan.py#L39) and [here](https://github.com/impira/docquery/blob/doc-classifier/src/docquery/cmd/scan.py#L68).

I think a good unit test in the future would be to go through and run each of the pipelines from (a) the CLI and (b) the python module and verify that the returned answers match, as well as verifying that both entry points haven't been inadvertently broken.

File: src/docquery/ext/pipeline_document_classification.py
Comment (jeffsrobertson) [inline src/docquery/ext/pipeline_document_classification.py:320]: If we want to extend the model to be able to predict "No value", this can be done pretty easily by adding an additional mask here `outputs > 0` (this corresponds to a sigmoid output of > 0.5, the typical threshold for binary classification). In other words, the results returned by `postprocess_standard()` would _only_ include the top_k predictions in which their respective logit was > 0.

I'm not sure how to extend this to account for when `len(model_outputs) > 1`, though. Something I'll have to think about for a bit

File: src/docquery/ext/pipeline_document_classification.py
Comment (jeffsrobertson) [inline src/docquery/ext/pipeline_document_classification.py:300]: Will this break if `top_k > len(answers)`?

**Review by jeffsrobertson — CHANGES_REQUESTED** [review]
Looks great :+1: I left plenty of comments, many of which are nits and can probably be ignored, but I think I also identified a few breaking bugs that need to be addressed before merging. In particular, `docquery/cmd/scan.py` requires a few very easy fixes in order to get the CLI working again with the `--classify` flag. I also have a couple items that I need to look further into on my end of this work, which I've left comments referencing.

👍 Let me know if you'd like me to clarify or expand on any of the comments I left!

File: src/docquery/ext/pipeline_document_classification.py
Comment (ankrgyl) [reply src/docquery/ext/pipeline_document_classification.py:320]: Okay makes sense. Maybe you could add this as a follow up PR? Filed here so we don't lose it: https://github.com/impira/docquery/issues/38

File: src/docquery/transformers_patch.py
Comment (ankrgyl) [reply src/docquery/transformers_patch.py:23]: Good catch

File: src/docquery/ext/pipeline_document_classification.py
Comment (jeffsrobertson) [reply src/docquery/ext/pipeline_document_classification.py:152]: 👍  That works too

File: src/docquery/ext/pipeline_document_classification.py
Comment (jeffsrobertson) [reply src/docquery/ext/pipeline_document_classification.py:213]: Ah okay I was mistaken on this. We use the Roberta fast tokenizer in the QA pipeline which is what I was thinking of. This is good as-is, sorry about that!

File: src/docquery/ext/pipeline_document_classification.py
Comment (jeffsrobertson) [reply src/docquery/ext/pipeline_document_classification.py:247]: For doc classification, the first token of the token sequence is the `[CLS]` token and the last one is the `[SEP]` token, with every token in between assigned `0` for its sequence id.

I would personally remove the dependence on sequence_ids entirely, as it's not really useful for classification, and write this if block like the following:

```
if i == self.tokenizer.cls_token_id:
    bbox.append(DEFAULT_CLS_BBOX)
elif i == self.tokenizer.sep_token_id:
    bbox.append(DEFAULT_SEP_BBOX)
elif i == self.tokenizer.pad_token_id:
    bbox.append(DEFAULT_PAD_BBOX)
else:
    bbox.append(boxes[w])
```

Note that the above code can also be used in the QA pipeline (although I believe an additional condition needs to be added to add the appropriate bboxes for the "question" part of the token sequence).

File: src/docquery/ext/pipeline_document_classification.py
Comment (jeffsrobertson) [reply src/docquery/ext/pipeline_document_classification.py:309]: 👍 Sounds good, happy to lead this change

**Comment by jeffsrobertson** [comment]
Below is a diff that applies a couple minor changes to `docquery/cmd/scan.py` to get the CLI working again. Could you include these changes in your PR?

```
diff --git a/src/docquery/cmd/scan.py b/src/docquery/cmd/scan.py
index 640ee9b..4e3223e 100644
--- a/src/docquery/cmd/scan.py
+++ b/src/docquery/cmd/scan.py
@@ -36,7 +36,7 @@ def build_parser(subparsers, parent_parser):
     parser.add_argument(
         "--classify-checkpoint",
         default=None,
-        help=f"A custom model checkpoint to use (other than {PIPELINE_DEFAULTS['image-classification']})",
+        help=f"A custom model checkpoint to use (other than {PIPELINE_DEFAULTS['document-classification']})",
     )
 
     parser.set_defaults(func=main)
@@ -65,7 +65,7 @@ def main(args):
     log.info("Ready to start evaluating!")
 
     if args.classify:
-        classify = pipeline("image-classification", model=args.classify_checkpoint)
+        classify = pipeline("document-classification", model=args.classify_checkpoint)
 
     max_fname_len = max(len(str(p)) for (p, d) in docs)
     max_question_len = max(len(q) for q in args.questions) if len(args.questions) > 0 else 0
@@ -73,7 +73,7 @@ def main(args):
         if i > 0 and len(args.questions) > 1:
             print("")
         if args.classify:
-            cls = classify(images=d.preview[0])[0]
+            cls = classify(image=d.preview[0])[0]
             print(f"{str(p):<{max_fname_len}} Document Type: {cls['label']}")
 
         for q in args.questions:
```

**Comment by jeffsrobertson** [comment]
Everything else looks good, thanks for the quick changes. After you've applied the above diff and had a chance to look at [this](https://github.com/impira/docquery/pull/37#discussion_r980246856) comment (I'll leave it up to you if you want to make this change or not though), I'm happy to give a ✅ !

**Comment by ankrgyl** [comment]
@jeffsrobertson just applied your changes -- please take a look when you have a chance.

**Review by jeffsrobertson — APPROVED** [review]
Awesome, everything looks good to me! You have my ✅

**Pull request merged**
